### PR TITLE
Narrowing Blob

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#11167 Narrowing Blob**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9623916/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11238 Some improvements to IValue&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9644700/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11258 Improve Tensor() constructor&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9652088/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11260 Fix intrusive_ptr move/copy for different NullType's&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9652089/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11294 Remove manual refcounting from Tensor class&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9663476/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11352 Remove intrusive_ptr::reclaim() in Storage&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9694327/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11353 Simplify union payload copying&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9694326/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11355 Simplify IValue::toTensor()&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9694918/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11402 Simplify Blob move constructor/assignment&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9728631/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11414 IValue can store Blob&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9731326/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11481 Blob doesn't allow access to destroyCall anymore&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/DD9756666/)

Narrow the Blob API as preparation for merging Blob/IValue

- get rid of templated IsType and Operator::InputIsType / OutputIsType
- Use 'using' instead of 'typedef' for DestroyCall (just for readability)

Differential Revision: [D9623916](https://our.internmc.facebook.com/intern/diff/DD9623916/)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Don't forget to say in your commit message what you actually narrowed. (In this case, you got rid of templated IsType and changed the deleter type... be more specific!)

File: caffe2/core/blob.h [hunk 1]
```diff
@@ -286,7 +280,7 @@ class CAFFE2_API Blob {
   }
   TypeMeta meta_;
   void* pointer_ = nullptr;
-  DestroyCall destroy_ = nullptr;
+  DestroyCall* destroy_ = nullptr;
```
Comment (ezyang) [inline caffe2/core/blob.h]: `const DestroyCall*`?

**Comment by ezyang** [comment]
Also, if a codemod was used, put it in the commit description too.

File: caffe2/core/blob.h (hunk 1 above)
Comment (smessmer) [reply caffe2/core/blob.h]: tbh, I'm not sure what a const function pointer would be semantically. It's not like a function object is mutable and you can change its implementation at runtime.

**Comment by smessmer** [comment]
I expanded the description. Didn't use a codemod but changed the call sites manually.

**Pull request closed without merging**